File: site/tools/generate-docs/rest-types.ts
import { Message } from './message';
import { JsonRestEnum, JsonRestObject, JsonRestTypes } from './types';

const { log } = console;

export class RestTypes {
  /**
   * A mapping from a message's full field name to the REST type
   * ex: frdrpcCloseRecommendationsResponse.considered_channels -> integer
   */
  fieldTypes = new Map<string, string>();
  /**
   * A mapping from an enum's name to the REST type
   * ex: ForceClosedChannelAnchorState -> string
   */
  enumTypes = new Map<string, string>();

  constructor(types: JsonRestTypes) {
    const entries = Object.entries(types);
    log(`Creating ${entries.length} RestTypes`);
    entries
      .filter(([name, schema]) => schema.type !== 'object')
      .forEach(([name, schema]) =>
        this.parseEnum(name, schema as JsonRestEnum)
      );
    entries
      .filter(([name, schema]) => schema.type === 'object')
      .forEach(([name, schema]) =>
        this.parseSchema(name, schema as JsonRestObject)
      );
  }

  parseEnum(typeName: string, schema: JsonRestEnum) {
    this.enumTypes.set(typeName, schema.type);
  }

  parseSchema(typeName: string, schema: JsonRestObject) {
    // skip if there are no properties defined
    if (!schema.properties) return;
    Object.entries(schema.properties).forEach(([fieldName, details]) => {
      const fullName = `${typeName}.${fieldName}`;
      let fieldType = 'unknown';
      if (details.$ref) {
        // $ref looks like: '#/definitions/InvoiceInvoiceState'
        // check if it is an enum or fallback to 'object'
        const refName = details.$ref.substring(
          details.$ref.lastIndexOf('/') + 1
        );
        const enumType = this.enumTypes.get(refName);
        fieldType = enumType || 'object';
      } else if (details.type) {
        fieldType = details.type;
      } else {
        throw new Error(`Could not determine the field type for ${fullName}`);
      }
      this.fieldTypes.set(fullName, fieldType);
    });
  }

  updateMessage(message: Message) {
    const msgRestType = this.convertGrpcTypeToRest(message.fullName);
    message.fields.forEach((field) => {
      const restType = `${msgRestType}.${field.name}`;
      if (this.fieldTypes.has(restType)) {
        field.restType = this.fieldTypes.get(restType);
      }
    });
  }

  /**
   * Converts the gRPC full type to a REST type
   * Examples:
   *   lnrpc.AddInvoiceResponse -> lnrpcAddInvoiceResponse
   */
  convertGrpcTypeToRest(typeName: string) {
    const parts = typeName.split('.');
    return `${parts[0]}${parts[parts.length - 1]}`;
  }
}

File: models/rest_types.go
package models

import (
	"fmt"
	"os"
	"strings"

	"github.com/lightninglabs/lightning-api-ng/defs"
)

type RestTypes struct {
	// FieldTypes is a mapping from a message's full field name to the REST
	// type.
	// ex: frdrpcCloseRecommendationsResponse.considered_channels -> integer
	FieldTypes map[string]string
	// EnumTypes is a mapping from an enum's name to the REST type
	// ex: ForceClosedChannelAnchorState -> string
	EnumTypes map[string]string
}

// NewRestTypes creates a new RestTypes from a map of RESTType definitions.
func NewRestTypes(restTypesDef map[string]*defs.RESTType) *RestTypes {
	fmt.Printf("Creating %d RestTypes\n", len(restTypesDef))
	r := &RestTypes{}
	r.FieldTypes = make(map[string]string)
	r.EnumTypes = make(map[string]string)

	// First parse all enums so we can reference them when parsing objects
	for name, restType := range restTypesDef {
		if restType.Type != "object" {
			r.parseEnum(name, restType)
		}
	}

	// Next, parse all objects.
	for name, restType := range restTypesDef {
		if restType.Type == "object" {
			r.parseObject(name, restType)
		}
	}
	return r
}

// parseEnum adds a RESTType definition to the EnumTypes map
func (r *RestTypes) parseEnum(typeName string, restType *defs.RESTType) {
	r.EnumTypes[typeName] = restType.Type
}

// parseObject parses a RESTType definition and adds it to the FieldTypes map
func (r *RestTypes) parseObject(typeName string, restType *defs.RESTType) {
	// skip if there are no properties defined
	if restType.Properties == nil {
		return
	}
	for fieldName, details := range restType.Properties {
		fullName := typeName + "." + fieldName
		r.FieldTypes[fullName] = details.Type
		fieldType := "unknown"
		if details.Ref != "" {
			// $ref looks like: '#/definitions/InvoiceInvoiceState'
			// check if it is an enum or fallback to 'object'
			index := strings.LastIndex(details.Ref, "/") + 1
			refName := details.Ref[index:]

			if enumType, ok := r.EnumTypes[refName]; ok {
				fieldType = enumType
			} else {
				fieldType = "object"
			}
		} else if details.Type != "" {
			fieldType = details.Type
		} else {
			fmt.Printf("Could not determine the field type "+
				"for %s\n", fullName)
			os.Exit(1)
		}
		r.FieldTypes[fullName] = fieldType
	}
}

// UpdateMessage updates a Message's fields with the REST type
func (r *RestTypes) UpdateMessage(msg *Message) {
	msgRestType := convertGrpcTypeToRest(msg.FullName)

	// update REST type and placement from restTypes
	for _, field := range msg.Fields {
		restTypeName := msgRestType + "." + field.Name
		if restType, ok := r.FieldTypes[restTypeName]; ok {
			field.RestType = restType
		}
	}
}

// convertGrpcTypeToRest converts a gRPC type to a REST type
// ex: lnrpc.AddInvoiceResponse -> lnrpcAddInvoiceResponse
func convertGrpcTypeToRest(typeName string) string {
	parts := strings.Split(typeName, ".")
	return parts[0] + parts[len(parts)-1]
}
